IntelliJ IDEA freezes when plugin is downloading an asset

<!-- 
    NOTE: Did you check the FAQs (https://one-dark.gitbook.io/jetbrains/help/faqs)
    to see if your problem already has a work around?
-->

**Describe the bug**

At random times, my IDE is freezing for 10-40 seconds. Last time it froze, here's what I found in the logs:

```
2021-03-16 17:27:07,343 [ 153518]   INFO - kelton.integrations.RestClient - Attempting to fetch remote asset: https://jetbrains.assets.unthrottled.io/one-dark/sentry-dsn.txt 
2021-03-16 17:27:47,479 [ 193654]   WARN - kelton.integrations.RestClient - Unable to get remote asset: https://jetbrains.assets.unthrottled.io/one-dark/sentry-dsn.txt for raisins 
org.apache.http.conn.ConnectTimeoutException: Connect to jetbrains.assets.unthrottled.io:443 [jetbrains.assets.unthrottled.io/99.86.159.84, jetbrains.assets.unthrottled.io/99.86.159.38, jetbrains.assets.unthrottled.io/99.86.159.106, jetbrains.assets.unthrottled.io/99.86.159.96, jetbrains.assets.unthrottled.io/2600:9000:2134:1e00:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:8c00:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:7600:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:fe00:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:1c00:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:1200:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:a00:c:b2bb:a240:93a1, jetbrains.assets.unthrottled.io/2600:9000:2134:c200:c:b2bb:a240:93a1] failed: connect timed out
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:151)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:376)
	at org.apache.http.impl.execchain.MinimalClientExec.execute(MinimalClientExec.java:186)
	at org.apache.http.impl.client.MinimalHttpClient.doExecute(MinimalHttpClient.java:108)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at com.markskelton.integrations.RestClient.performGet(RestClient.kt:20)
	at com.markskelton.integrations.ErrorReporter$Companion$1.configure(ErrorReporter.kt:36)
	at io.sentry.Sentry.init(Sentry.java:132)
	at io.sentry.Sentry.init(Sentry.java:119)
	at com.markskelton.integrations.ErrorReporter.<clinit>(ErrorReporter.kt:34)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:684)
	at com.intellij.openapi.extensions.impl.ExtensionComponentAdapter.instantiateClass(ExtensionComponentAdapter.java:51)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter$SimpleConstructorInjectionAdapter.instantiateClass(XmlExtensionAdapter.java:133)
	at com.intellij.openapi.extensions.impl.ExtensionComponentAdapter.createInstance(ExtensionComponentAdapter.java:43)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter.createInstance(XmlExtensionAdapter.java:69)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.processAdapter(ExtensionPointImpl.java:501)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.processAdapters(ExtensionPointImpl.java:435)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.calcExtensionList(ExtensionPointImpl.java:246)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.getExtensionList(ExtensionPointImpl.java:240)
	at com.intellij.openapi.extensions.ExtensionPointName.getExtensionList(ExtensionPointName.java:42)
	at com.intellij.diagnostic.IdeErrorsDialog.getSubmitter(IdeErrorsDialog.java:923)
	at com.intellij.diagnostic.IdeErrorsDialog.getSubmitter(IdeErrorsDialog.java:913)
	at com.intellij.diagnostic.DefaultIdeaErrorLogger.canHandle(DefaultIdeaErrorLogger.java:55)
	at com.intellij.diagnostic.DialogAppender.appendToLoggers(DialogAppender.java:86)
	at com.intellij.diagnostic.DialogAppender.lambda$queueAppend$0(DialogAppender.java:59)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:976)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:843)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:454)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:773)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:453)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:822)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:507)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:368)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	... 49 more
2021-03-16 17:27:47,534 [ 193709]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 40413ms, details saved to /Users/taademaq/Library/Logs/JetBrains/IntelliJIdea2020.3/threadDumps-freeze-20210316-172712-IU-203.7717.56-PlainSocketImpl.socketConnect-40sec 
```
It seems to indicate the plugin is the source of the freeze, maybe something is done in the wrong thread?

**To Reproduce**
<!-- Steps to reproduce the behavior -->
Now it's hard for me to say why the connection times out and how to reproduce. But I'm behind a corporate proxy with TLS interception and other shenanigans, so I think it might be related. IntelliJ is configured to autodetect the proxy using a PAC script.

**Expected Behavior**
<!-- A clear and concise description of what you expected to happen. -->
Even if a connection is slow or unavailable, the plugin should not freeze the IDE.

**Environment**

```
IntelliJ IDEA 2020.3.3 (Ultimate Edition)
Build #IU-203.7717.56, built on March 15, 2021
...
Runtime version: 11.0.10+8-b1145.96 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 10.15.7
GC: ParNew, ConcurrentMarkSweep
Memory: 1088M
Cores: 8
Non-Bundled Plugins: Key Promoter X, com.markskelton.one-dark-theme, mobi.hsz.idea.gitignore, String Manipulation, com.jetbrains.codeWithMe, org.jetbrains.kotlin, com.viartemev.requestmapper, com.intellij.lang.jsgraphql
```

Plugin version: 5.1.5

**Side issue**

The log says `raisins` instead of `reason` I think 😄

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,8 @@
 
 #### Unreleased
 
+- Fixed UI Freeze issues
+
 #### 5.1.5
 
 - Added identifier under caret error stripe color.

--- a/src/main/kotlin/com/markskelton/integrations/ErrorReporter.kt
+++ b/src/main/kotlin/com/markskelton/integrations/ErrorReporter.kt
@@ -3,6 +3,7 @@ package com.markskelton.integrations
 import com.intellij.ide.IdeBundle
 import com.intellij.ide.plugins.PluginManagerCore
 import com.intellij.ide.ui.LafManager
+import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.application.ApplicationNamesInfo
 import com.intellij.openapi.application.ex.ApplicationInfoEx
 import com.intellij.openapi.application.impl.ApplicationInfoImpl
@@ -16,7 +17,10 @@ import com.intellij.util.Consumer
 import com.intellij.util.text.DateFormatUtil
 import com.markskelton.settings.ThemeSettings
 import com.markskelton.tools.runSafelyWithResult
-import io.sentry.*
+import io.sentry.Sentry
+import io.sentry.SentryEvent
+import io.sentry.SentryLevel
+import io.sentry.SentryOptions
 import io.sentry.protocol.Message
 import io.sentry.protocol.User
 import java.awt.Component
@@ -28,16 +32,19 @@ import java.util.stream.Collectors
 
 class ErrorReporter : ErrorReportSubmitter() {
   override fun getReportActionText(): String = "Report Anonymously"
+
   companion object {
 
     init {
-      Sentry.init { options: SentryOptions ->
-        options.dsn =
-          RestClient.performGet(
-            "https://jetbrains.assets.unthrottled.io/one-dark/sentry-dsn.txt"
-          )
-            .map { it.trim() }
-            .orElse("https://cb598170e51a44adbf0079abe2d79624@o403546.ingest.sentry.io/5267019?maxmessagelength=50000")
+      ApplicationManager.getApplication().executeOnPooledThread {
+        Sentry.init { options: SentryOptions ->
+          options.dsn =
+            RestClient.performGet(
+              "https://jetbrains.assets.unthrottled.io/one-dark/sentry-dsn.txt"
+            )
+              .map { it.trim() }
+              .orElse("https://cb598170e51a44adbf0079abe2d79624@o403546.ingest.sentry.io/5267019?maxmessagelength=50000")
+        }
       }
 
       Sentry.setUser(
@@ -55,21 +62,23 @@ class ErrorReporter : ErrorReportSubmitter() {
     consumer: Consumer<in SubmittedReportInfo>
   ): Boolean {
     return runSafelyWithResult({
-      events.forEach {
-        Sentry.captureEvent(
-          addSystemInfo(
-            SentryEvent()
-              .apply {
-                this.level = SentryLevel.ERROR
-                this.serverName = getAppName().second
-                this.setExtra("Additional Info", additionalInfo ?: "None")
+      ApplicationManager.getApplication().executeOnPooledThread {
+        events.forEach {
+          Sentry.captureEvent(
+            addSystemInfo(
+              SentryEvent()
+                .apply {
+                  this.level = SentryLevel.ERROR
+                  this.serverName = getAppName().second
+                  this.setExtra("Additional Info", additionalInfo ?: "None")
+                }
+            ).apply {
+              this.message = Message().apply {
+                this.message = it.throwableText
               }
-          ).apply {
-            this.message = Message().apply {
-              this.message = it.throwableText
             }
-          }
-        )
+          )
+        }
       }
       true
     }) {

--- a/src/main/kotlin/com/markskelton/notification/Notifications.kt
+++ b/src/main/kotlin/com/markskelton/notification/Notifications.kt
@@ -13,9 +13,7 @@ import org.intellij.lang.annotations.Language
 val UPDATE_MESSAGE: String = """
       What's New?<br>
       <ul>
-        <li>Added identifier under caret error stripe color.</li>
-        <li>Themed Default Inlays</li>
-        <li>Update Ruby's local variable color</li>
+        <li>Fixed UI Freezing issue.</li>
       </ul>
       <br>Please see the <a href="https://github.com/one-dark/jetbrains-one-dark-theme/blob/master/CHANGELOG.md">Changelog</a> for more details.
       <br>

